File: python/ch17_agent/agent_ch17/features/skills.py
"""第七章按需加载 Skill。

Java 对照：`SkillRegistry` 类似一个只读的配置/路由注册表。启动时只读取
每个 `SKILL.md` 的 frontmatter，得到名称和描述；模型真正调用 `load_skill`
时才读取正文。这样 System Prompt 不会在启动时塞入所有技能说明。
"""

from __future__ import annotations

import re
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml

from ..adapters.filesystem import is_windows_reserved_component
from ..core.tools import ToolContext, ToolDefinition, ToolResult, tool_error, tool_success

LOAD_SKILL_TOOL_NAME = "load_skill"
DEFAULT_SKILLS_DIRECTORY = "skills"
DEFAULT_MAX_CATALOG_ENTRIES = 100
DEFAULT_MAX_CATALOG_BYTES = 8_000
MAX_SKILL_NAME_LENGTH = 64
SKILL_NAME_PATTERN = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")


class SkillError(Exception):
    """Skill 领域错误的共同父类。"""


class SkillPathError(SkillError):
    """Skill 根目录、Skill 目录或 manifest 逃出了受控边界。"""


class SkillManifestError(SkillError):
    """SKILL.md 的 frontmatter 缺失、YAML 错误或字段不符合契约。"""


class DuplicateSkillError(SkillError):
    """多个目录声明了同一个 Skill 名称。"""


class SkillNameError(SkillError):
    """请求或 manifest 中的 Skill 名称不合法。"""


class SkillNotFoundError(SkillError):
    """名称格式正确，但当前注册表没有该 Skill。"""


@dataclass(frozen=True, slots=True)
class SkillSummary:
    """公开给模型的目录条目，只包含路由所需的两项元数据。"""

    name: str  # 稳定的工具路由名称，也必须等于目录名。
    description: str  # 一行路由说明，不包含 Skill 私有正文。


@dataclass(frozen=True, slots=True)
class _SkillRecord:
    """注册表内部记录；路径保存逻辑入口，加载时会重新解析真实路径。"""

    summary: SkillSummary  # 扫描阶段校验过的名称和描述。
    directory_name: str  # workspace/skills 下的目录名。
    directory_path: Path  # 逻辑目录入口，防止扫描后替换链接不被发现。
    manifest_path: Path  # 逻辑 SKILL.md 入口，加载时重新做 realpath 校验。


def _validate_skill_name(name: str) -> None:
    """校验名称到目录的映射，拒绝路径穿越和 Windows 设备名。"""
    if (
        not isinstance(name, str)
        or not 1 <= len(name) <= MAX_SKILL_NAME_LENGTH
        or SKILL_NAME_PATTERN.fullmatch(name) is None
        or is_windows_reserved_component(name)
    ):
        raise SkillNameError(f"Skill 名称不合法: {name}")


def _validate_skill_input(value: Mapping[str, Any]) -> bool:
    """`load_skill` 只允许一个非空、格式安全的 name 字段。"""
    if set(value) != {"name"} or not isinstance(value.get("name"), str):
        return False
    try:
        _validate_skill_name(value["name"])
    except SkillNameError:
        return False
    return True


def _workspace_root(workspace: str) -> Path:
    """取得 workspace 的真实目录，类似 Java 中的受控根目录解析器。"""
    try:
        root = Path(workspace).resolve(strict=True)
    except OSError as error:
        raise SkillPathError("工作区不存在或无法解析") from error
    if not root.is_dir():
        raise SkillPathError("工作区不是目录")
    return root


def _is_inside(root: Path, candidate: Path) -> bool:
    """用 Path.relative_to 判断包含关系，避免字符串前缀误判。"""
    try:
        candidate.relative_to(root)
        return True
    except ValueError:
        return False


def _checked_real_directory(path: Path, root: Path, message: str) -> Path:
    """realpath 后确认目标仍是 root 内的真实目录。"""
    try:
        physical = path.resolve(strict=True)
    except OSError as error:
        raise SkillPathError(message) from error
    if not physical.is_dir() or not _is_inside(root, physical):
        raise SkillPathError(message)
    return physical


def _checked_real_file(path: Path, root: Path, message: str) -> Path:
    """realpath 后确认目标仍是 root 内的真实文件。"""
    try:
        physical = path.resolve(strict=True)
    except OSError as error:
        raise SkillPathError(message) from error
    if not physical.is_file() or not _is_inside(root, physical):
        raise SkillPathError(message)
    return physical


def _resolve_skill_root(workspace_root: Path, skills_directory: str) -> Path:
    """只允许 workspace 内的相对 skills 目录。"""
    if not isinstance(skills_directory, str) or not skills_directory or "\x00" in skills_directory:
        raise SkillPathError("Skills 目录必须是非空相对路径")
    normalized = skills_directory.replace("\\", "/")
    if normalized.startswith("/") or re.fullmatch(r"[A-Za-z]:.*", normalized):
        raise SkillPathError("Skills 目录必须是相对 workspace 的路径")
    parts = [part for part in normalized.split("/") if part not in {"", "."}]
    if not parts or ".." in parts:
        raise SkillPathError("Skills 目录不能包含父目录片段 ..")
    if any(is_windows_reserved_component(part) for part in parts):
        raise SkillPathError("Skills 目录包含 Windows 保留路径组件")
    target = workspace_root.joinpath(*parts)
    if target.exists() and not _is_inside(workspace_root, target.resolve()):
        raise SkillPathError("Skills 目录逃逸 workspace")
    return target


def _read_frontmatter(path: Path) -> str:
    """只返回 frontmatter 字节对应的文本，不解码正文。

    这里先读 bytes，再在第二个独立的 `---` 行处截断；因此正文即使含有非法
    UTF-8，扫描阶段也不会失败，真正加载正文时才会严格解码。
    """
    selected: list[bytes] = []
    with path.open("rb") as manifest:
        for line in manifest:
            selected.append(line)
            if len(selected) > 1 and line.rstrip(b"\r\n") == b"---":
                break
    try:
        return b"".join(selected).decode("utf-8")
    except UnicodeDecodeError as error:
        raise SkillManifestError("Skill manifest 的 frontmatter 不是合法 UTF-8") from error


def _parse_document(text: str, source: Path) -> tuple[SkillSummary, str]:
    """解析完整 SKILL.md，返回元数据和 frontmatter 后的原始正文。"""
    lines = text.splitlines(keepends=True)
    if not lines or lines[0].rstrip("\r\n") != "---":
        raise SkillManifestError("SKILL.md 必须以 YAML frontmatter 开始")
    closing_index = next(
        (index for index, line in enumerate(lines[1:], start=1) if line.rstrip("\r\n") == "---"),
        None,
    )
    if closing_index is None:
        raise SkillManifestError("SKILL.md 缺少 frontmatter 结束分隔符")
    try:
        metadata = yaml.safe_load("".join(lines[1:closing_index]))
    except yaml.YAMLError as error:
        raise SkillManifestError("Skill frontmatter 不是合法 YAML") from error
    if not isinstance(metadata, dict):
        raise SkillManifestError("Skill frontmatter 必须是对象")
    name = metadata.get("name")
    description = metadata.get("description")
    if not isinstance(name, str) or not isinstance(description, str):
        raise SkillManifestError("Skill frontmatter 必须包含字符串 name 和 description")
    try:
        _validate_skill_name(name)
    except SkillNameError as error:
        raise SkillManifestError("Skill frontmatter 中的 name 不合法") from error
    normalized_description = description.strip()
    if (
        not normalized_description
        or "\n" in normalized_description
        or "\r" in normalized_description
    ):
        raise SkillManifestError("Skill description 必须是非空单行文本")
    return SkillSummary(name, normalized_description), "".join(lines[closing_index + 1 :])


def _bounded_catalog(
    records: list[_SkillRecord], max_entries: int, max_bytes: int
) -> tuple[SkillSummary, ...]:
    """按条目数和 UTF-8 字节预算截断，但不截断半条目录行。"""
    result: list[SkillSummary] = []
    used = 0
    for record in sorted(records, key=lambda item: item.summary.name):
        if len(result) >= max_entries:
            break
        line = f"- **{record.summary.name}**: {record.summary.description}"
        entry_bytes = len(line.encode("utf-8")) + (1 if result else 0)
        if used + entry_bytes > max_bytes:
            break
        result.append(record.summary)
        used += entry_bytes
    return tuple(result)


class SkillRegistry:
    """绑定一个 workspace 的 Skill 元数据注册表和 `load_skill` 工具。"""

    def __init__(
        self,
        workspace_root: Path,
        skills_root: Path,
        records: dict[str, _SkillRecord],
        catalog_entries: tuple[SkillSummary, ...],
    ) -> None:
        self._workspace_root = workspace_root  # 所有 Skill 路径的最高信任边界。
        self._skills_root = skills_root  # workspace 下的 skills 根目录。
        self._records = dict(records)  # 内部完整记录，模型只能看到摘要。
        self.names = tuple(sorted(records))  # 稳定名称快照，便于测试和日志。
        self.catalog_entries = catalog_entries  # 已应用预算的公开摘要快照。
        self.tool_definition = ToolDefinition(
            LOAD_SKILL_TOOL_NAME,
            "加载当前 workspace catalog 中指定 Skill 的完整说明。",
            {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "minLength": 1,
                        "maxLength": MAX_SKILL_NAME_LENGTH,
                        "pattern": SKILL_NAME_PATTERN.pattern,
                    }
                },
                "required": ["name"],
                "additionalProperties": False,
            },
            "read",
            self._handle_load,
            _validate_skill_input,
        )

    @classmethod
    def scan(
        cls,
        workspace: str,
        *,
        skills_directory: str = DEFAULT_SKILLS_DIRECTORY,
        max_catalog_entries: int = DEFAULT_MAX_CATALOG_ENTRIES,
        max_catalog_bytes: int = DEFAULT_MAX_CATALOG_BYTES,
    ) -> SkillRegistry:
        """扫描一级 Skill 目录，只解析 frontmatter 并建立不可变摘要。"""
        if (
            isinstance(max_catalog_entries, bool)
            or not isinstance(max_catalog_entries, int)
            or max_catalog_entries <= 0
        ):
            raise ValueError("max_catalog_entries 必须是正整数")
        if (
            isinstance(max_catalog_bytes, bool)
            or not isinstance(max_catalog_bytes, int)
            or max_catalog_bytes <= 0
        ):
            raise ValueError("max_catalog_bytes 必须是正整数")
        workspace_root = _workspace_root(workspace)
        skills_path = _resolve_skill_root(workspace_root, skills_directory)
        if not skills_path.exists():
            return cls(workspace_root, skills_path, {}, ())
        skills_root = _checked_real_directory(
            skills_path, workspace_root, "Skills 目录逃逸 workspace"
        )
        discovered: list[_SkillRecord] = []
        records: dict[str, _SkillRecord] = {}
        for entry in sorted(skills_root.iterdir(), key=lambda item: item.name):
            if not entry.is_dir() and not entry.is_symlink():
                continue
            directory_real = _checked_real_directory(entry, skills_root, "Skill 目录逃逸 Skills 根")
            manifest = entry / "SKILL.md"
            if not manifest.exists():
                continue
            manifest_real = _checked_real_file(manifest, directory_real, "Skill manifest 逃逸目录")
            summary, _body = _parse_document(_read_frontmatter(manifest_real), manifest_real)
            if summary.name in records:
                raise DuplicateSkillError(f"Skill 名称重复: {summary.name}")
            record = _SkillRecord(summary, entry.name, entry, manifest)
            records[summary.name] = record
            discovered.append(record)
        for record in discovered:
            if record.summary.name != record.directory_name:
                raise SkillManifestError("Skill name 必须与目录名一致")
        catalog = _bounded_catalog(list(records.values()), max_catalog_entries, max_catalog_bytes)
        return cls(workspace_root, skills_root, records, catalog)

    def render_catalog(self) -> str:
        """渲染只含名称和描述的目录，不读取正文。"""
        return "\n".join(
            f"- **{entry.name}**: {entry.description}" for entry in self.catalog_entries
        )

    def load_skill(self, name: str) -> str:
        """重新检查路径并返回 frontmatter 后的正文。"""
        _validate_skill_name(name)
        record = self._records.get(name)
        if record is None:
            raise SkillNotFoundError(f"Skill 不存在: {name}")
        current_root = _checked_real_directory(
            self._skills_root, self._workspace_root, "Skills 目录已逃逸 workspace"
        )
        current_directory = _checked_real_directory(
            record.directory_path, current_root, "Skill 目录已逃逸 Skills 根"
        )
        current_manifest = _checked_real_file(
            record.manifest_path, current_directory, "Skill manifest 已逃逸目录"
        )
        try:
            document = current_manifest.read_bytes().decode("utf-8")
        except UnicodeDecodeError as error:
            raise SkillManifestError("Skill manifest 不是合法 UTF-8") from error
        summary, body = _parse_document(document, current_manifest)
        if summary.name != record.summary.name or summary.name != record.directory_name:
            raise SkillManifestError("Skill name 必须与目录名一致")
        return body

    def _handle_load(self, arguments: Mapping[str, Any], context: ToolContext) -> ToolResult:
        """校验 ToolContext 属于本注册表后，把领域错误映射成稳定工具结果。"""
        try:
            context_workspace = _workspace_root(context.workspace)
        except SkillPathError:
            return tool_error("skill_workspace_error", "当前工作区无法解析")
        if context_workspace != self._workspace_root:
            return tool_error("skill_workspace_mismatch", "Skill 注册表属于另一个工作区")
        try:
            return tool_success(self.load_skill(str(arguments["name"])))
        except SkillNotFoundError:
            return tool_error("skill_not_found", "请求的 Skill 未注册")
        except SkillPathError:
            return tool_error("skill_path_escape", "已注册 Skill 的路径不再安全")
        except SkillManifestError:
            return tool_error("invalid_skill", "已注册 Skill 的清单无效")
        except Exception:  # noqa: BLE001
            return tool_error("skill_load_error", "Skill 无法加载")

File: typescript/chapters/ch18/src/features/skills.ts
// Skill 管理：扫描技能目录、解析 frontmatter 并做路径白名单校验，为模型提供 load_skill 的 catalog 与内容读取。
import {
  closeSync,
  existsSync,
  lstatSync,
  openSync,
  readSync,
  readdirSync,
  realpathSync,
  statSync,
} from "node:fs";
import {
  readFile as readFileBytes,
  realpath as realpathAsync,
  stat as statAsync,
} from "node:fs/promises";
import { isAbsolute, relative, resolve, sep, win32 } from "node:path";
import { TextDecoder } from "node:util";

import { parse } from "yaml";
import { z } from "zod";

import { isWindowsReservedComponent } from "../core/filesystem.js";
import type { ToolContext, ToolDefinition, ToolResult } from "../core/tools.js";
import { toolError, toolSuccess } from "../core/tools.js";

// 技能目录、工具名与 catalog 上限是 Skill 功能的公共边界。
export const LOAD_SKILL_TOOL_NAME = "load_skill";
export const DEFAULT_SKILLS_DIRECTORY = "skills";
export const DEFAULT_MAX_CATALOG_ENTRIES = 100;
export const DEFAULT_MAX_CATALOG_BYTES = 8_000;
export const MAX_SKILL_NAME_LENGTH = 64;

// Skill 注册表只公开受限目录中的清单摘要，正文必须通过 load_skill 按需读取。
const UTF8_DECODER = new TextDecoder("utf-8", { fatal: true });
const SKILL_NAME_REGEXP = /^[a-z0-9]+(?:-[a-z0-9]+)*$/u;

// 校验名称为字母数字带连字符的合法标识，并拒绝 Windows 设备名。
const loadSkillInputSchema = z
  .object({
    name: z
      .string()
      .min(1)
      .max(MAX_SKILL_NAME_LENGTH)
      .regex(SKILL_NAME_REGEXP)
      .refine((name) => !isWindowsReservedComponent(name))
      .describe("The exact Skill name from the available catalog."),
  })
  .strict();

export type LoadSkillInput = Readonly<z.output<typeof loadSkillInputSchema>>;

export class SkillError extends Error {}

export class SkillPathError extends SkillError {}

export class SkillManifestError extends SkillError {}

export class DuplicateSkillError extends SkillError {}

export class SkillNameError extends SkillError {}

export class SkillNotFoundError extends SkillError {}

export interface SkillSummary {
  readonly name: string;
  readonly description: string;
}

interface SkillRecord extends SkillSummary {
  readonly directoryName: string;
  readonly directoryPath: string;
  readonly manifestPath: string;
}

export interface SkillScanOptions {
  readonly skillsDirectory?: string;
  readonly maxCatalogEntries?: number;
  readonly maxCatalogBytes?: number;
}

interface ParsedSkillDocument extends SkillSummary {
  readonly body: string;
}

// 扫描生成不可变元数据快照；加载时重新校验每一层真实路径。
export class SkillRegistry {
  // 扫描时记录入口，加载时再次校验真实路径，以防扫描后链接被替换。
  readonly #workspaceRoot: string;
  readonly #skillsRoot: string;
  readonly #records: ReadonlyMap<string, SkillRecord>;
  readonly names: readonly string[];
  readonly catalogEntries: readonly SkillSummary[];
  readonly toolDefinition: ToolDefinition<LoadSkillInput>;

  private constructor(
    workspaceRoot: string,
    skillsRoot: string,
    records: ReadonlyMap<string, SkillRecord>,
    catalogEntries: readonly SkillSummary[],
  ) {
    this.#workspaceRoot = workspaceRoot;
    this.#skillsRoot = skillsRoot;
    this.#records = new Map(records);
    this.names = Object.freeze([...records.keys()].sort());
    this.catalogEntries = Object.freeze(catalogEntries.map((entry) => Object.freeze({ ...entry })));
    this.toolDefinition = Object.freeze({
      name: LOAD_SKILL_TOOL_NAME,
      description: "Load the full instructions for one Skill listed in the workspace catalog.",
      inputSchema: loadSkillInputSchema,
      effect: "read",
      handler: (input: LoadSkillInput, context: ToolContext) => this.#handleLoad(input, context),
    });
  }

  static scan(workspace: string, options: SkillScanOptions = {}): SkillRegistry {
    // Catalog 有条目和字节上限，避免 workspace 文件把系统提示词无限扩张。
    const workspaceRoot = resolveWorkspace(workspace);
    const skillsDirectory =
      options.skillsDirectory === undefined ? DEFAULT_SKILLS_DIRECTORY : options.skillsDirectory;
    const maxCatalogEntries = positiveInteger(
      options.maxCatalogEntries,
      DEFAULT_MAX_CATALOG_ENTRIES,
      "maxCatalogEntries",
    );
    const maxCatalogBytes = positiveInteger(
      options.maxCatalogBytes,
      DEFAULT_MAX_CATALOG_BYTES,
      "maxCatalogBytes",
    );
    const skillsPath = resolveSkillRoot(workspaceRoot, skillsDirectory);
    if (!existsSync(skillsPath)) {
      return new SkillRegistry(workspaceRoot, skillsPath, new Map(), []);
    }

    const skillsRoot = checkedRealDirectory(
      skillsPath,
      workspaceRoot,
      `Skills directory escapes workspace: ${skillsDirectory}`,
    );
    const discovered: SkillRecord[] = [];
    const byName = new Map<string, SkillRecord>();

    for (const entry of readdirSync(skillsRoot, { withFileTypes: true }).sort((left, right) =>
      compareSkillNames(left.name, right.name),
    )) {
      const lexicalDirectory = resolve(skillsRoot, entry.name);
      const information = lstatSync(lexicalDirectory);
      if (!information.isDirectory() && !information.isSymbolicLink()) {
        continue;
      }
      const directoryPath = checkedRealDirectory(
        lexicalDirectory,
        skillsRoot,
        `Skill directory escapes Skills root: ${entry.name}`,
      );
      const lexicalManifest = resolve(directoryPath, "SKILL.md");
      if (!existsSync(lexicalManifest)) {
        continue;
      }
      const manifestPath = checkedRealFile(
        lexicalManifest,
        directoryPath,
        `Skill manifest escapes its directory: ${entry.name}`,
      );
      const parsed = parseSkillDocument(readFrontmatter(manifestPath), manifestPath);
      // 保存逻辑入口；显式加载时重新解析，才能发现扫描后的链接替换。
      const record: SkillRecord = Object.freeze({
        name: parsed.name,
        description: parsed.description,
        directoryName: entry.name,
        directoryPath: lexicalDirectory,
        manifestPath: resolve(lexicalDirectory, "SKILL.md"),
      });
      if (byName.has(record.name)) {
        throw new DuplicateSkillError(`Duplicate Skill name: ${record.name}`);
      }
      byName.set(record.name, record);
      discovered.push(record);
    }

    for (const record of discovered) {
      if (record.name !== record.directoryName) {
        throw new SkillManifestError(
          `Skill name must match its directory: ${record.directoryName}`,
        );
      }
    }

    const ordered = [...byName.values()].sort((left, right) =>
      compareSkillNames(left.name, right.name),
    );
    const catalog = boundedCatalog(ordered, maxCatalogEntries, maxCatalogBytes);
    return new SkillRegistry(
      workspaceRoot,
      skillsRoot,
      new Map(ordered.map((record) => [record.name, record])),
      catalog,
    );
  }

  renderCatalog(): string {
    return this.catalogEntries
      .map((entry) => `- **${entry.name}**: ${entry.description}`)
      .join("\n");
  }

  async loadSkill(name: string): Promise<string> {
    // 读取请求边界重新解析 workspace、Skill 目录和 manifest 的物理位置。
    validateSkillName(name);
    const record = this.#records.get(name);
    if (record === undefined) {
      throw new SkillNotFoundError(`Skill not found: ${name}`);
    }

    // load_skill 位于请求路径，使用异步 I/O，并重新检查每一层真实路径。
    const currentSkillsRoot = await checkedRealDirectoryAsync(
      this.#skillsRoot,
      this.#workspaceRoot,
      "Skills directory escapes workspace",
    );
    const currentDirectory = await checkedRealDirectoryAsync(
      record.directoryPath,
      currentSkillsRoot,
      `Skill directory escapes Skills root: ${name}`,
    );
    const currentManifest = await checkedRealFileAsync(
      record.manifestPath,
      currentDirectory,
      `Skill manifest escapes its directory: ${name}`,
    );
    const document = parseSkillDocument(
      decodeUtf8(await readFileBytes(currentManifest), currentManifest),
      currentManifest,
    );
    if (document.name !== record.name || document.name !== record.directoryName) {
      throw new SkillManifestError(`Skill name must match its directory: ${name}`);
    }
    return document.body;
  }

  // handler 先验证 workspace 一致性，再通过 loadSkill 读取并返回正文。
  async #handleLoad(input: LoadSkillInput, context: ToolContext): Promise<ToolResult> {
    let contextWorkspace: string;
    try {
      contextWorkspace = await resolveWorkspaceAsync(context.workspace);
    } catch (error) {
      if (error instanceof SkillPathError) {
        return toolError("skill_workspace_error", "Current workspace could not be resolved");
      }
      throw error;
    }
    if (contextWorkspace !== this.#workspaceRoot) {
      return toolError(
        "skill_workspace_mismatch",
        "Skill registry belongs to a different workspace",
      );
    }
    try {
      return toolSuccess(await this.loadSkill(input.name));
    } catch (error) {
      if (error instanceof SkillNotFoundError) {
        return toolError("skill_not_found", "Requested Skill is not registered");
      }
      if (error instanceof SkillPathError) {
        return toolError("skill_path_escape", "Registered Skill path is no longer safe");
      }
      if (error instanceof SkillManifestError) {
        return toolError("invalid_skill", "Registered Skill manifest is invalid");
      }
      return toolError("skill_load_error", "Skill could not be loaded");
    }
  }
}

function positiveInteger(value: number | undefined, fallback: number, label: string): number {
  const selected = value === undefined ? fallback : value;
  if (!Number.isInteger(selected) || selected <= 0) {
    throw new RangeError(`${label} must be a positive integer`);
  }
  return selected;
}

function resolveWorkspace(workspace: string): string {
  let root: string;
  try {
    root = realpathSync.native(workspace);
  } catch {
    throw new SkillPathError(`Workspace does not exist: ${workspace}`);
  }
  if (!statSync(root).isDirectory()) {
    throw new SkillPathError(`Workspace is not a directory: ${workspace}`);
  }
  return root;
}

async function resolveWorkspaceAsync(workspace: string): Promise<string> {
  let root: string;
  try {
    root = await realpathAsync(workspace);
  } catch {
    throw new SkillPathError(`Workspace does not exist: ${workspace}`);
  }
  if (!(await statAsync(root)).isDirectory()) {
    throw new SkillPathError(`Workspace is not a directory: ${workspace}`);
  }
  return root;
}

function resolveSkillRoot(workspaceRoot: string, skillsDirectory: string): string {
  if (skillsDirectory.length === 0 || skillsDirectory.includes("\0")) {
    throw new SkillPathError("Skills directory must be a non-empty relative path");
  }
  const normalized = skillsDirectory.replaceAll("\\", "/");
  if (
    isAbsolute(skillsDirectory) ||
    win32.isAbsolute(skillsDirectory) ||
    /^[A-Za-z]:/u.test(skillsDirectory) ||
    normalized.startsWith("/")
  ) {
    throw new SkillPathError("Skills directory must be relative to the workspace");
  }
  const parts = normalized.split("/").filter((part) => part.length > 0 && part !== ".");
  if (parts.length === 0 || parts.includes("..")) {
    throw new SkillPathError("Skills directory must not contain parent segments");
  }
  for (const part of parts) {
    if (isWindowsReservedComponent(part)) {
      throw new SkillPathError(`Skills directory contains a reserved path component: ${part}`);
    }
  }
  const target = resolve(workspaceRoot, ...parts);
  if (!isInside(workspaceRoot, target)) {
    throw new SkillPathError("Skills directory escapes workspace");
  }
  return target;
}

function checkedRealDirectory(path: string, root: string, message: string): string {
  let physical: string;
  try {
    physical = realpathSync.native(path);
  } catch {
    throw new SkillPathError(message);
  }
  if (!isInside(root, physical) || !statSync(physical).isDirectory()) {
    throw new SkillPathError(message);
  }
  return physical;
}

function checkedRealFile(path: string, root: string, message: string): string {
  let physical: string;
  try {
    physical = realpathSync.native(path);
  } catch {
    throw new SkillPathError(message);
  }
  if (!isInside(root, physical) || !statSync(physical).isFile()) {
    throw new SkillPathError(message);
  }
  return physical;
}

async function checkedRealDirectoryAsync(
  path: string,
  root: string,
  message: string,
): Promise<string> {
  let physical: string;
  try {
    physical = await realpathAsync(path);
  } catch {
    throw new SkillPathError(message);
  }
  if (!isInside(root, physical) || !(await statAsync(physical)).isDirectory()) {
    throw new SkillPathError(message);
  }
  return physical;
}

async function checkedRealFileAsync(path: string, root: string, message: string): Promise<string> {
  let physical: string;
  try {
    physical = await realpathAsync(path);
  } catch {
    throw new SkillPathError(message);
  }
  if (!isInside(root, physical) || !(await statAsync(physical)).isFile()) {
    throw new SkillPathError(message);
  }
  return physical;
}

function isInside(root: string, candidate: string): boolean {
  const child = relative(root, candidate);
  return (
    child.length === 0 || (child !== ".." && !child.startsWith(`..${sep}`) && !isAbsolute(child))
  );
}

// 只读取 frontmatter 结束分隔符之前的字节，避免扫描时解码全量正文。
function readFrontmatter(path: string): string {
  const descriptor = openSync(path, "r");
  const completeLines: Buffer[] = [];
  let pending = Buffer.alloc(0);
  const chunk = Buffer.allocUnsafe(4_096);
  try {
    while (true) {
      const count = readSync(descriptor, chunk, 0, chunk.byteLength, null);
      if (count === 0) {
        break;
      }
      let offset = 0;
      while (offset < count) {
        const newline = chunk.indexOf(0x0a, offset);
        if (newline === -1 || newline >= count) {
          pending = Buffer.concat([pending, chunk.subarray(offset, count)]);
          break;
        }
        const line = Buffer.concat([pending, chunk.subarray(offset, newline + 1)]);
        pending = Buffer.alloc(0);
        completeLines.push(line);
        if (completeLines.length > 1 && isFrontmatterSeparator(line)) {
          // catalog 只解码 frontmatter；正文留到 load_skill 被显式调用之后。
          return decodeUtf8(Buffer.concat(completeLines), path);
        }
        offset = newline + 1;
      }
    }
    if (pending.byteLength > 0) {
      completeLines.push(pending);
    }
    return decodeUtf8(Buffer.concat(completeLines), path);
  } finally {
    closeSync(descriptor);
  }
}

function isFrontmatterSeparator(line: Buffer): boolean {
  let end = line.byteLength;
  if (end > 0 && line[end - 1] === 0x0a) {
    end -= 1;
  }
  if (end > 0 && line[end - 1] === 0x0d) {
    end -= 1;
  }
  return end === 3 && line[0] === 0x2d && line[1] === 0x2d && line[2] === 0x2d;
}

function decodeUtf8(bytes: Uint8Array, source: string): string {
  try {
    return UTF8_DECODER.decode(bytes);
  } catch {
    throw new SkillManifestError(`Skill manifest is not valid UTF-8: ${source}`);
  }
}

function splitLinesKeepEnds(text: string): string[] {
  const lines: string[] = [];
  let start = 0;
  for (let index = 0; index < text.length; index += 1) {
    if (text[index] === "\n") {
      lines.push(text.slice(start, index + 1));
      start = index + 1;
    }
  }
  if (start < text.length) {
    lines.push(text.slice(start));
  }
  return lines;
}

function stripLineEnding(line: string): string {
  if (line.endsWith("\r\n")) {
    return line.slice(0, -2);
  }
  return line.endsWith("\n") ? line.slice(0, -1) : line;
}

// 校验 frontmatter 格式与结构，确保 name、description 合规。
function parseSkillDocument(text: string, source: string): ParsedSkillDocument {
  const lines = splitLinesKeepEnds(text);
  if (lines.length === 0 || stripLineEnding(lines[0] as string) !== "---") {
    throw new SkillManifestError(`Skill manifest must begin with YAML frontmatter: ${source}`);
  }
  const closingIndex = lines.findIndex(
    (line, index) => index > 0 && stripLineEnding(line) === "---",
  );
  if (closingIndex === -1) {
    throw new SkillManifestError(`Skill manifest has no closing frontmatter delimiter: ${source}`);
  }

  let metadata: unknown;
  try {
    metadata = parse(lines.slice(1, closingIndex).join(""));
  } catch {
    throw new SkillManifestError(`Skill frontmatter is not valid YAML: ${source}`);
  }
  if (typeof metadata !== "object" || metadata === null || Array.isArray(metadata)) {
    throw new SkillManifestError(`Skill frontmatter must be a mapping: ${source}`);
  }
  const name = Reflect.get(metadata, "name");
  const description = Reflect.get(metadata, "description");
  if (typeof name !== "string" || typeof description !== "string") {
    throw new SkillManifestError(
      `Skill frontmatter requires string name and description: ${source}`,
    );
  }
  try {
    validateSkillName(name);
  } catch (error) {
    if (error instanceof SkillNameError) {
      throw new SkillManifestError(`Skill frontmatter contains an invalid name: ${source}`);
    }
    throw error;
  }
  const normalizedDescription = description.trim();
  if (
    normalizedDescription.length === 0 ||
    normalizedDescription.includes("\n") ||
    normalizedDescription.includes("\r")
  ) {
    throw new SkillManifestError(`Skill description must be one non-empty line: ${source}`);
  }
  return Object.freeze({
    name,
    description: normalizedDescription,
    body: lines.slice(closingIndex + 1).join(""),
  });
}

function validateSkillName(name: string): void {
  if (
    name.length > MAX_SKILL_NAME_LENGTH ||
    !SKILL_NAME_REGEXP.test(name) ||
    isWindowsReservedComponent(name)
  ) {
    throw new SkillNameError(`Invalid Skill name: ${name}`);
  }
}

function compareSkillNames(left: string, right: string): number {
  if (left === right) {
    return 0;
  }
  return left < right ? -1 : 1;
}

// Catalog 按条目数和 UTF-8 byte 双重上限截断，返回不可变快照。
function boundedCatalog(
  records: readonly SkillRecord[],
  maxEntries: number,
  maxBytes: number,
): readonly SkillSummary[] {
  const catalog: SkillSummary[] = [];
  let usedBytes = 0;
  for (const record of records) {
    if (catalog.length >= maxEntries) {
      break;
    }
    const line = `- **${record.name}**: ${record.description}`;
    const separatorBytes = catalog.length === 0 ? 0 : 1;
    const entryBytes = Buffer.byteLength(line, "utf8") + separatorBytes;
    if (usedBytes + entryBytes > maxBytes) {
      break;
    }
    catalog.push(Object.freeze({ name: record.name, description: record.description }));
    usedBytes += entryBytes;
  }
  return Object.freeze(catalog);
}
